### PR TITLE
Remove all usages of unstable rust features

### PR DESCRIPTION
**Pull request by Frizi**
### Pull Request Description

Removed all `#![feature]` flags, except for `#![feature(test)]`. Once parser benchmarks are ported to something that is compatible with stable rust, we will be able to switch to it.

### CODE REVIEW

File: build/ci_utils/src/extensions/command.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ pub trait CommandExt {
             let _ = write!(ret, "\n\twith working directory: {}", cwd.display());
         };
         let env = self.as_std().get_envs();
-        if !env.is_empty() {
+        if env.len() != 0 {
```
Comment (farmaazon) [inline build/ci_utils/src/extensions/command.rs:18]: When I was using rust last time, every structure with `len` was strongly encouraged by linter to have also `is_emptry`. Has something changed?

File: lib/rust/parser/src/macros/built_in.rs [hunk 2]
```diff
@@ -401,25 +401,20 @@ fn try_parse_doc_comment<'s>(
...
+        Some(Item::Token(token @ Token { variant: token::Variant::TextStart(_), .. })) if token.code == "##"
+    ) {
         let Some(syntax::Tree {
-            variant:
-                syntax::tree::Variant::Documented(box syntax::tree::Documented {
-                    mut documentation,
-                    expression: None,
-                    ..
-                }),
+            variant: syntax::tree::Variant::Documented(mut documented),
```
Comment (farmaazon) [inline lib/rust/parser/src/macros/built_in.rs:409]: I think we lost the check if `expression` is None.

**Review by farmaazon — APPROVED** [review]

File: build/ci_utils/src/extensions/command.rs (hunk 1 above)
Comment (Frizi) [reply build/ci_utils/src/extensions/command.rs:18]: That `is_empty` was unstable, part of `exact_size_is_empty` feature.

File: lib/rust/parser/src/macros/built_in.rs (hunk 2 above)
Comment (Frizi) [reply lib/rust/parser/src/macros/built_in.rs:409]: The alternative is unreachable, so I'm not sure it matters in this case. @kazcw could you check if that match was indeed significant here?

File: build/ci_utils/src/extensions/command.rs (hunk 1 above)
Comment (farmaazon) [reply build/ci_utils/src/extensions/command.rs:18]: So, rust does not follow its own rules :)

File: lib/rust/parser/src/syntax/tree/block.rs [hunk 3]
```diff
@@ -210,18 +207,17 @@ fn to_operator_block_expression<'s>(
...
-        && properties.can_form_section()
-    {
-        let expression = precedence.resolve_offset(1, &mut items).unwrap();
-        let Some(Item::Token(operator)) = items.pop() else { unreachable!() };
-        let operator = Ok(operator.with_variant(token::variant::Operator()));
-        Ok(OperatorBlockExpression { operator, expression })
-    } else {
-        Err(precedence.resolve(&mut items).unwrap())
+    match &items[..] {
+        [Item::Token(a), b, ..]
```
Comment (kazcw) [inline lib/rust/parser/src/syntax/tree/block.rs:211]: This is nice! I don't use slice patterns enough.

File: lib/rust/parser/src/syntax/tree/block.rs [hunk 4]
```diff
@@ -210,18 +207,17 @@ fn to_operator_block_expression<'s>(
...
+            if b.left_visible_offset().width_in_spaces != 0
+                && a.operator_properties().map_or(false, |p| p.can_form_section()) =>
```
Comment (kazcw) [inline lib/rust/parser/src/syntax/tree/block.rs]: I think `is_some_and` is clearer for this.

File: lib/rust/parser/src/macros/built_in.rs (hunk 2 above)
Comment (kazcw) [reply lib/rust/parser/src/macros/built_in.rs:409]: Let's replace it with a debug assertion. This checks a precondition for the function that, if not satisfied, would result in dropping tokens. If tokens are dropped, the sooner we catch it the easier it is to debug.

**Review by kazcw — APPROVED** [review]

File: build/ci_utils/src/extensions/command.rs (hunk 1 above)
Comment (Frizi) [reply build/ci_utils/src/extensions/command.rs:18]: It's not an official rust rule, but a clippy configuration. Also it only checks things that actually have stable `is_empty` method, so this change doesn't trigger a warning.

**Review by hubertp — APPROVED** [review]

File: lib/rust/parser/debug/src/bin/lexer.rs [hunk 5]
```diff
@@ -1,9 +1,5 @@
 //! Run the lexer from the command line, for understanding the early stages of the parser.
 
-// === Features ===
-#![feature(exact_size_is_empty)]
-#![feature(let_chains)]
-#![feature(if_let_guard)]
```
Comment (JaroslavTulach) [inline lib/rust/parser/debug/src/bin/lexer.rs:6]: Does removing some of these feature have an impact on the size of generated `.so` binary?

**Review by JaroslavTulach — APPROVED** [review]

File: lib/rust/parser/debug/src/bin/lexer.rs (hunk 5 above)
Comment (kazcw) [reply lib/rust/parser/debug/src/bin/lexer.rs:6]: Some of the switches might be a little less optimized, but it won't have a noticeable impact.

**Pull request merged**